lib/util/bufferedPassThrough.ts

```diff
@@ -0,0 +1,186 @@
+
+type Encoding = BufferEncoding | 'buffer';
+
+type ChunkData = Buffer | string;
+
+type Chunk = {
+  chunk: ChunkData;
+  encoding: Encoding;
+};
+
+type Callback = (error?: Error) => void;
+
+/**
+ * This streams accumulate chunks data into a buffer until the amount of data is equal or exceed  the buffer size.
+ * Then, it emits a single chunk with the accumulated data.
+ * 
+ * This stream is useful when you want to reduce the amount of chunk produced by a stream.
+ * This can helps reducing the number of syscall made by the stream consumer.
+ */
+export class BufferedPassThrough extends stream.Duplex {
+  private bufferSize: number;
+  private buffer: Buffer;
+  private offset: number;
+
+  constructor (options: stream.DuplexOptions = { highWaterMark: 8196 }) {
+    super(options);
+    this.bufferSize = options.highWaterMark;
+    this.buffer = Buffer.alloc(options.highWaterMark);
+    this.offset = 0; 
+  }
+
+  /**
+   * Writes a string or buffer to the internal buffer starting at the internal buffer offset.
+   * Data will be copied from the given start to the end position.
+   * 
+   * @param data 
+   * @param start Where to start copying/writing data from
+   * @param end Where to end copying/writing data from
+   * @param encoding Type of encoding to use
+   * @returns How many bytes have been copied / written to the internal buffer.
+   */
+  private writeToBuffer (data: ChunkData, start: number, end: number, encoding: Encoding): number {
+    if (encoding === 'buffer') {
+      return (data as Buffer).copy(this.buffer, this.offset, start, end);
+    }
+    if (start === 0 && end === data.length) {
+      return this.buffer.write(data as string, this.offset, end - start, encoding);
+    }
+    return this.buffer.write((data as string).slice(start, end), this.offset, end - start, encoding);
+  }
+
+  private async writeChunks (chunks: Chunk[]): Promise<void> {
+    for (const chunk of chunks) {
+      await this.writeChunkData(chunk.chunk, chunk.encoding);
+    }
+  }
+
+  /**
+   * Writes a chunk of data to the internal buffer.
+   * If the internal buffer is full, it will be pushed to the stream.
+   * @param chunk 
+   * @param encoding 
+   */
+  private async writeChunkData (chunk: ChunkData, encoding: Encoding): Promise<void> {
+
+    let chunkOffset = 0;
+
+    while (chunkOffset < chunk.length) {
+      
+      const remainingChunkSize = chunk.length - chunkOffset;
+      const remainingBufferSize = this.bufferSize - this.offset;
+
+      // If the remaining bytes in the chunk exceed the remaining bytes in the internal buffer
+      if (remainingChunkSize >= remainingBufferSize) {
+        // Write as much as the remaining bytes as possible to the buffer  from t
+        chunkOffset += this.writeToBuffer(chunk, chunkOffset, chunkOffset + remainingBufferSize, encoding);
+
+        // Sends the whole buffer to the stream since it is full
+        if (! this.push(this.buffer)) {
+          await new Promise((res) => {
```

Comment:
We should add the [arrow-parens](https://eslint.org/docs/rules/arrow-parens) rule to eslint for this one

```suggestion
          await new Promise(res => {
```